fix: record an external gate's own depth, not its skipped decomposition's (#359)
qBraid/pyqasm: CHANGELOG.md

### Removed

### Fixed

> - Fixed external and verbatim-box gates counting the depth of the decomposition they skipped: `unroll(external_gates=["crz"])` on a single `crz` reported `depth() == 12` while emitting one statement. An external gate now records its own depth, like the custom-gate path already did. ([#352](https://github.com/qBraid/pyqasm/issues/352))

Reviewer: **Type:** Maintenance
**Severity:** Low

**Rationale:** "like the custom-gate path already did" reads as though the custom-gate path is correct in general. It is not, for nested external custom gates — the same user-visible symptom this entry describes still occurs there.

Evidence, identical on base and on this PR (so **pre-existing and not a regression from this change**):

```qasm
gate inner a, b { crz(0.5) a, b; }
gate outer a, b { inner a, b; h a; }
qubit[2] q;
outer q[0], q[1];
```

With `external_gates=["outer"]` this emits exactly one statement, `outer q[0], q[1];`, but reports `depth() == 13`. With `external_gates=["inner", "outer"]` it reports `2` for the same single statement. A single-level external custom gate correctly reports `1`, which is why the existing tests pass.

The cause is precisely the pattern this PR improved on. `_visit_custom_gate_operation` assigns the flag unconditionally at visitor.py:1285 and clears it unconditionally at visitor.py:1317-1318, with no save-restore. The inner custom gate therefore clobbers the outer gate's flag, and the outer gate never records its own depth. The new code here avoids exactly that by saving `prev_recording` and restoring it in a `finally` — applying the same pattern at :1285 would close the remaining gap.

**Change Requested:** Soften the wording so the entry does not claim the custom-gate path is already fully correct. Consider a follow-up issue for the nested case; fixing it here would widen the PR beyond #352.

```suggestion
- Fixed external and verbatim-box gates counting the depth of the decomposition they skipped: `unroll(external_gates=["crz"])` on a single `crz` reported `depth() == 12` while emitting one statement. An external gate now records its own depth, matching how a single-level external custom gate is already handled. ([#352](https://github.com/qBraid/pyqasm/issues/352))
```

Author: Confirmed your repro exactly: `external_gates=["outer"]` reports `depth() == 13`, `["inner", "outer"]` reports `2`, and a single-level external custom gate reports `1` — identical on base and on this branch, so pre-existing rather than a regression here.

Took your wording in ecd3cf5. Agreed the fix belongs in a separate PR: `_visit_custom_gate_operation` needs the same save-restore around the flag at :1285/:1317 that this PR uses. Happy to open a follow-up issue for it unless you would rather file it with your repro.
